# Hermitian Matrices, Unitary Matrices and Tensor Product


# Understanding Complex Hermitian and Unitary Matrices

This tutorial delves into two pivotal concepts in quantum computing and linear algebra: complex Hermitian and Unitary matrices. Both play critical roles in quantum mechanics and quantum computing, underpinning the behavior of quantum systems and the operations performed on them.

## Objectives
- Define and explore the properties of complex Hermitian matrices.
- Understand and implement complex Unitary matrices in Python.
- Highlight their significance and applications in quantum computing.

Prepare to enhance your mathematical toolkit with these foundational concepts, illustrated with Python examples for a hands-on learning experience.



## Complex Hermitian Matrices

A Hermitian matrix, or self-adjoint matrix, is a square matrix that is equal to its own conjugate transpose. This property makes it a natural representation of observable quantities in quantum mechanics.

### Properties
- The eigenvalues of a Hermitian matrix are always real.
- The eigenvectors of a Hermitian matrix corresponding to different eigenvalues are orthogonal.
- Hermitian matrices are extensively used in quantum mechanics to describe physical observables.

### Python Example: Defining a Hermitian Matrix


In [2]:

import numpy as np

# Define a complex Hermitian matrix
hermitian_matrix = np.array([[2+0j, 2-1j], [2+1j, 3+0j]])

# Check if the matrix is Hermitian
is_hermitian = np.allclose(hermitian_matrix, hermitian_matrix.conj().T)

is_hermitian, hermitian_matrix


(True,
 array([[2.+0.j, 2.-1.j],
        [2.+1.j, 3.+0.j]]))


## Complex Unitary Matrices

A Unitary matrix is a square matrix whose conjugate transpose is also its inverse. Unitary matrices preserve the inner product, making them essential for describing quantum gates and evolution in quantum computing.

### Properties
- Preserves the norm of vectors, hence preserving quantum states during transformations.
- The determinant of a Unitary matrix has an absolute value of 1.
- Used to represent reversible quantum operations and time evolution in quantum systems.

### Python Example: Defining a Unitary Matrix


In [2]:

# Define a complex Unitary matrix
unitary_matrix = np.array([[1/np.sqrt(2), 1/np.sqrt(2)], [1/np.sqrt(2)*1j, -1/np.sqrt(2)*1j]])

# Check if the matrix is Unitary
# np.eye(2) returns the identity of size 2x2
is_unitary = np.allclose(np.dot(unitary_matrix, unitary_matrix.conj().T), np.eye(2))

is_unitary, unitary_matrix


(True,
 array([[ 0.70710678+0.j        ,  0.70710678+0.j        ],
        [ 0.        +0.70710678j, -0.        -0.70710678j]]))

# Tensor product


## Tensor Product for Complex Vectors

The tensor product of two complex vectors creates a new vector in a higher-dimensional space, allowing us to combine quantum states in quantum computing.

### Example


In [ ]:

import numpy as np

# Define two complex vectors for the tensor product
v1 = np.array([1+2j, 3+4j])
v2 = np.array([5+6j, 7+8j])

# Calculate the tensor product
tensor_product_v = np.kron(v1, v2)

tensor_product_v



## Tensor Product for Complex Matrices

The tensor product for matrices combines individual matrices into a larger matrix, crucial for representing multiple qubit systems in quantum computing.

### Example


In [ ]:

# Define two complex matrices for the tensor product
M1 = np.array([[1+2j, 3+4j], [5+6j, 7+8j]])
M2 = np.array([[9+10j, 11+12j], [13+14j, 15+16j]])

# Calculate the tensor product
tensor_product_m = np.kron(M1, M2)

tensor_product_m


# Exercises on Complex Vector and Matrix Operations

This set of exercises is designed to test your understanding of various concepts related to complex vector and matrix operations, which are foundational in quantum computing. Each exercise presents a concrete case for you to apply what you've learned about Hermitian matrices, Unitary matrices, and tensor products.

NOTE: VERIFY YOUR COMPUTATIONS IN PAPER AND IN THE COMPUTER.

## Exercise 1: Complex Hermitian Matrices

Consider the matrix:

$$ H = \begin{bmatrix} 3 & 2+i \\ 2-i & 1 \end{bmatrix} $$

- Verify if $ H $ is a Hermitian matrix.
- If it is, find its eigenvalues.

## Exercise 2: Complex Unitary Matrices

Consider the matrix:

$$ U = \frac{1}{\sqrt{2}}\begin{bmatrix} 1 & i \\ i & 1 \end{bmatrix} $$

- Verify if $ U $ is a Unitary matrix.
- Compute $ UU^\dagger $ to confirm its Unitarity, where $ U^\dagger $ denotes the conjugate transpose of $ U $.

## Exercise 3: Tensor Product for Complex Vectors

Given the complex vectors:

$$ \mathbf{v} = \begin{bmatrix} 1+i \\ 2-i \end{bmatrix}, \quad \mathbf{w} = \begin{bmatrix} 1-2i \\ 3 \end{bmatrix} $$

Calculate the tensor product $ \mathbf{v} \otimes \mathbf{w} $.

## Exercise 4: Tensor Product for Complex Matrices

Given the matrices:

$$ M_1 = \begin{bmatrix} 0 & 1 \\ 1 & 0 \end{bmatrix}, \quad M_2 = \begin{bmatrix} i & 0 \\ 0 & -i \end{bmatrix} $$

Calculate the tensor product $ M_1 \otimes M_2 $.

## Exercise 5: Modelling quantum computations with vectors and matrices

Using matrices and vectors, implement a model of the Mach/Zehnder interferometer.

![Mach-Zehnder interferometer](images/Mach-Zehnder-Interferometer.png)

## Exercise 6: Composing quantum systems 

Using matrices and vectors, implement a model of the following circuit.

![Mach-Zehnder interferometer](images/Deutsch-Algorithm.png)

Use the following MAtrix for $U_f$:

![Mach-Zehnder interferometer](images/ExampleUf.png)



## Respuestas ##

## Exercise 1: Complex Hermitian Matrices

Consider the matrix:

$$ H = \begin{bmatrix} 3 & 2+i \\ 2-i & 1 \end{bmatrix} $$

- Verify if $ H $ is a Hermitian matrix.
- If it is, find its eigenvalues.

### Respuesta 
como se ve en la salida del codigo la respuesta correcta es:

Si es una Hermitian matrix
    
$$ m = \begin{bmatrix} 3 & 2 + i \\ 2 - i & 1 \end{bmatrix} $$ 


$$ Eigenvalues = \begin{bmatrix} 4.4 - 6.7i & -0.4 + 1.1i  \end{bmatrix} $$ 

In [3]:
# Define a complex Hermitian matrix
hermitian_matrix = np.array([[3, 2+1j], [2-1j, 1]])

# Check if the matrix is Hermitian
is_hermitian = np.allclose(hermitian_matrix, hermitian_matrix.conj().T)

is_hermitian, hermitian_matrix



# Define your complex matrix
A = np.array([[3, 2+1j],
              [2-1j, 1+0j]], dtype=complex)

# Compute eigenvalues and eigenvectors
eigenvalues, eigenvectors = np.linalg.eig(A)

# Print the results
print("Eigenvalues:")
print(eigenvalues)


Eigenvalues:
[ 4.44948974-6.70989255e-17j -0.44948974+1.15877743e-17j]


## Exercise 2: Complex Unitary Matrices

Consider the matrix:

$$ U = \frac{1}{\sqrt{2}}\begin{bmatrix} 1 & i \\ i & 1 \end{bmatrix} $$

- Verify if $ U $ is a Unitary matrix.
- Compute $ UU^\dagger $ to confirm its Unitarity, where $ U^\dagger $ denotes the conjugate transpose of $ U $.

### Respuesta 
como se ve en la salida del codigo la respuesta correcta es:

True

In [7]:
# Define a complex Unitary matrix
unitary_matrix = np.array([[1, 1j], [1j, 1]])
unitary_matrix2 = unitary_matrix *1/np.sqrt(2)

# Check if the matrix is Unitary
is_unitary = np.allclose(np.dot(unitary_matrix2, unitary_matrix2.conj().T), np.eye(2))

is_unitary, unitary_matrix

(True,
 array([[1.+0.j, 0.+1.j],
        [0.+1.j, 1.+0.j]]))

## Exercise 3: Tensor Product for Complex Vectors

Given the complex vectors:

$$ \mathbf{v} = \begin{bmatrix} 1+i \\ 2-i \end{bmatrix}, \quad \mathbf{w} = \begin{bmatrix} 1-2i \\ 3 \end{bmatrix} $$

Calculate the tensor product $ \mathbf{v} \otimes \mathbf{w} $.

### Respuesta 
como se ve en la salida del codigo la respuesta correcta es:

$$ \mathbf{v} = \begin{bmatrix} 2-4i \\ 6 \\ 6-3i\end{bmatrix}$$

In [9]:
# Define two complex vectors for the tensor product
v1 = np.array([1+1, 2-1j])
v2 = np.array([1-2j, 3+0j])

# Calculate the tensor product
tensor_product_v = np.kron(v1, v2)

tensor_product_v


array([2.-4.j, 6.+0.j, 0.-5.j, 6.-3.j])

## Exercise 4: Tensor Product for Complex Matrices

Given the matrices:

$$ M_1 = \begin{bmatrix} 0 & 1 \\ 1 & 0 \end{bmatrix}, \quad M_2 = \begin{bmatrix} i & 0 \\ 0 & -i \end{bmatrix} $$

Calculate the tensor product $ M_1 \otimes M_2 $.

$$
M_1 = \begin{bmatrix}
    0 & 0 & i & 0 \\
    0 & 0 & 0 & -i \\
    i & 0 & 0 & 0 \\
    0 & -i & 0 & 0 \\
\end{bmatrix}
$$

In [3]:
import numpy as np

# Define the matrices M1 and M2
M1 = np.array([[0, 1],
               [1, 0]])

M2 = np.array([[1j, 0],
               [0, -1j]])

# Compute the tensor product M1 ⊗ M2
tensor_product = np.kron(M1, M2)

print(tensor_product)


[[0.+0.j 0.+0.j 0.+1.j 0.+0.j]
 [0.+0.j 0.-0.j 0.+0.j 0.-1.j]
 [0.+1.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.-1.j 0.+0.j 0.-0.j]]


## Exercise 5: Modelling quantum computations with vectors and matrices

Using matrices and vectors, implement a model of the Mach/Zehnder interferometer.

![Mach-Zehnder interferometer](images/Mach-Zehnder-Interferometer.png)

In [5]:
# Define the beam splitter matrix
def beam_splitter():
    return 1/np.sqrt(2) * np.array([[1, 1j],
                                     [1j, 1]])

# Define the phase shifter matrix with phase phi
def phase_shifter(phi):
    return np.array([[np.exp(1j * phi), 0],
                     [0, np.exp(1j * phi)]])

# Define the Mach-Zehnder interferometer matrix
def mach_zehnder_interferometer(phi):
    BS = beam_splitter()
    PS = phase_shifter(phi)
    
    # Construct the matrix representing the entire interferometer
    MZ_interferometer = np.dot(PS, np.dot(BS, np.dot(PS, BS)))
    
    return MZ_interferometer

# Input state vector representing the photons entering the interferometer
input_state = np.array([1, 0])  # Assuming input state is |0>, i.e., photon enters upper path

# Set the phase difference introduced by the phase shifter
phi = np.pi / 4  # Example: pi/4 phase difference

# Apply the Mach-Zehnder interferometer matrix to the input state
output_state = np.dot(mach_zehnder_interferometer(phi), input_state)

# Print the output state vector
print("Output state vector after passing through the Mach-Zehnder interferometer:")
print(output_state)


Output state vector after passing through the Mach-Zehnder interferometer:
[ 0.+0.j -1.+0.j]


## Exercise 6: Composing quantum systems 

Using matrices and vectors, implement a model of the following circuit.

![Mach-Zehnder interferometer](images/Deutsch-Algorithm.png)

Use the following MAtrix for $U_f$:

![Mach-Zehnder interferometer](images/ExampleUf.png)

### Respuesta 
como se ve en la salida del codigo la respuesta correcta es:
    
$$ Resultado = \begin{bmatrix} 0 - 0.70710678  \\ 0.70710678   + 0i \end{bmatrix} $$

                        

In [15]:
# Define two complex vectors for the tensor product
v1 = np.array([0, 1])
v2 = np.array([1, 0])

# Calculate the tensor product
tensor_product_v = np.kron(v1, v2)

# Define two complex matrices for the tensor product
M1 = np.array([[1, 1], [1, -1]])
M2 = np.array([[1, 0], [0, 1]])
M3 = 1/np.sqrt(2) * M1
# Calculate the tensor product
tensor_product_m = np.kron(M3, M2)

# matriz por vector 
product_tensor = np.dot(tensor_product_m, tensor_product_v)

#Matriz Uf por el resultado anterior
U = np.array([[0,1,0,0], [1,0,0,0], [0,0,1,0], [0,0,0,1]])

k =np.dot(U, product_tensor)

# Define two complex matrices for the tensor product
M4 = np.array([[1, 1], [1, -1]])
M5 = np.array([[1, 1], [1, -1]])
M6 = 1/np.sqrt(2) * M4
M7 = 1/np.sqrt(2) * M5

# Calculate the tensor product
tensor_product_l = np.kron(M6, M7)

#Resulatdo final
final = np.dot(tensor_product_l, k)

print("resultado =", final)


resultado = [ 0.         -0.70710678  0.70710678  0.        ]
